### PR TITLE
[serverless] Add lifecycle API endpoints

### PR DESCRIPTION
**Pull request by nhinsch**
### What does this PR do?

Add a new lifecycle API to the Serverless agent. The API consists of two endpoints: `/lambda/start-invocation` and `/lambda/end-invocation`. Together, these endpoints can replace the proxy in contexts where the proxy does not work.

- Rename the `ProxyProcessor` type to `LifecycleProcessor`, as it is now used by both the proxy and the lifecycle API
- Instantiate the `LifecycleProcessor` as a member of the `daemon`. Afterwards, it can be used by both the proxy and the lifecycle API.
- Consolidate all API routes in the daemon into a separate file, `routes.go`
- Move the `InvocationProcessor`, `InvocationStartDetails`, and `InvocationEndDetails` types from the `proxy` package to the `lifecycle` package as they are not specific to the proxy
- Add the new API endpoints that trigger the `LifecycleProcessor` without using the proxy

### Describe how to test/QA your changes

- Unit tests of the new endpoints are included in the PR
- In a future PR, we will add integration tests that cover using these endpoints
- The endpoints have been manually tested in a Lambda environment

### Reviewer's Checklist
- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.

### CODE REVIEW

File: cmd/serverless/main.go [hunk 1]
```diff
@@ -297,16 +297,20 @@ func runAgent(stopCh chan struct{}) (serverlessDaemon *daemon.Daemon, err error)
 
 	wg.Wait()
 
-	// start the proxy if needed
+	// set up invocation processor in the serverless Daemon to be used for the proxy and/or lifecycle API
+	serverlessDaemon.InvocationProcessor = &invocationlifecycle.LifecycleProcessor{
+		ExtraTags:           serverlessDaemon.ExtraTags,
+		MetricChannel:       serverlessDaemon.MetricAgent.GetMetricChannel(),
+		ProcessTrace:        serverlessDaemon.TraceAgent.Get().Process,
```
Comment (maxday) [inline cmd/serverless/main.go:304]: are we sure that `.MetricAgent` and `.TraceAgent` and `.TraceAgent.Get()` won't be nil at this point?

Comment (maxday) [inline cmd/serverless/main.go:304]: are we sure that `.MetricAgent` and `.TraceAgent` and `.TraceAgent.Get()` won't be nil at this point?

File: pkg/serverless/daemon/routes_test.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	"github.com/DataDog/datadog-agent/pkg/serverless/invocationlifecycle"
+	"github.com/stretchr/testify/assert"
+)
+
+type mockLifecycleProcessor struct {
+	OnInvokeStartCalled bool
+	OnInvokeEndCalled   bool
+}
+
+func (m *mockLifecycleProcessor) OnInvokeStart(*invocationlifecycle.InvocationStartDetails) {
```
Comment (maxday) [inline pkg/serverless/daemon/routes_test.go:23]: :100:

File: pkg/serverless/daemon/routes.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+// the invocation lifecyle feature without the use of the proxy.
+type StartInvocation struct {
+	daemon *Daemon
+}
+
+func (s *StartInvocation) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	log.Debug("Hit on the serverless.StartInvocation route.")
+
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
```
Comment (maxday) [inline pkg/serverless/daemon/routes.go:53]: Maybe we should add a unit test for this error

File: pkg/serverless/daemon/routes.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+// the invocation lifecyle feature without the use of the proxy.
+type EndInvocation struct {
+	daemon *Daemon
+}
+
+func (e *EndInvocation) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	log.Debug("Hit on the serverless.EndInvocation route.")
+
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
```
Comment (maxday) [inline pkg/serverless/daemon/routes.go:80]: same here?

**Review by maxday — APPROVED** [review]
I really like the new structure 💯  🎉

File: cmd/serverless/main.go (hunk 1 above)
Comment (nhinsch) [reply cmd/serverless/main.go:304]: I believe so. The MetricAgent is started on line 259, and there is a WaitGroup to ensure that the TraceAgent is started before we continue onwards and set the InvocationProcessor.

**Pull request merged**